### ISSUE TITLE
Screenshots not taken on failure when running tests in parallel with junit5

### ISSUE THREAD

**Issue by c-fau** [serenity-bdd/serenity-core#3334]
### What happened?

I created a junit5 parallel test runner following this [documentation](https://serenity-bdd.github.io/docs/tutorials/running_cucumber_scenarios_in_parallel) to run my cucumber feature files in parallel. However when serenity.take.screenshots=FOR_FAILURES screenshots are not taken when the step fails. If this value is set to any of the other values screenshots are taken and attached to the serenity report.

Screenshots are taken correctly and as expected when running the same test with a test runner using the @RunWith(CucumberWithSerenity.class) annotation.
By inserting breakpoints in BaseStepListener.class around the methods that take screenshots i've noticed that when using the parallel runner takeEndOfStepScreenshotForPlayback appears to be called, but when using the single threaded runner takeEndOfStepScreenshotFor is called instead. Not sure if that's helpful at all.

### What did you expect to happen?

Screenshots to be taken on failure of step

### Serenity BDD version

4.0.21

### JDK version

Correto 17.0.9.8.1

### Execution environment

MacOS X 14.1.1
Gradle 8.0
Chrome Firefox (same behaviour across both)

### How to reproduce the bug.

follow documentation below to create parallel test runner
https://serenity-bdd.github.io/docs/tutorials/running_cucumber_scenarios_in_parallel

create feature file with step that fails
generate aggregate report
observe no screenshot is taken

### How can we make it happen?

Add it to the Serenity BDD backlog and wait for a volunteer to pick it up

**Comment by wakaleo** [comment]
I can't reproduce this:

serenity.conf:
```hocon
serenity.take.screenshots=FOR_FAILURES
```

<img width="1385" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/55986/732c724a-b242-4866-914c-845e496c0c29">

<img width="1392" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/55986/ecc364cf-61b4-42e2-aeed-22cac76f04cf">

**Comment by c-fau** [comment]
Thanks for the prompt reply. I've chucked together a quick repo matching my set-up and I can confirm when running
`./gradlew clean test aggregate`
the generated report does not contain screenshots for the failures. serenity.take.screenshots is set to FOR_FAILURES in serenity.conf and I can confirm that when set to FOR_EACH action the screenshots are taken.


**FOR_FAILURES**
<img width="313" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/fbb49b20-bda7-40c0-9504-3c5b33a3343d">

<img width="1710" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/32946df5-7190-4800-ac3b-af8371231c25">




**FOR_EACH_STEP**
<img width="321" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/be2fe216-e6ef-4818-971a-21764b0e6a45">

<img width="1714" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/d8aade98-2684-464c-b84b-127810d13ba7">

**Comment by wakaleo** [comment]
Can you provide a sample project reproducing the issue?

**Comment by c-fau** [comment]
Apologies forgot to ling the project: https://github.com/c-fau/serenityCucumberTest
Just to add when ran with `./gradlew clean singleCucumber aggregate` and set to FOR_FAILURES, works as expected
<img width="1719" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/b444dbc1-350a-478e-92e1-ce9990c0ed3a">

**Comment by wakaleo** [comment]
Indeed, I can reproduce it with Gradle. The screenshot is recorded but does not appear in the reports. No idea why it would behave like that though. If you want to investigate on your side that might be a good place to start.

**Comment by c-fau** [comment]
Are you able to see the screenshot in target/site/serenity before of after aggregate has ran? I am unable to see the .png before of after.
Presumably then the screenshot is taken but not stored?

**Comment by wakaleo** [comment]
I can see the screenshot in target/site/serenity

**Comment by c-fau** [comment]
Hmm, that muddies things a bit. I cannot see the screenshot in that location, i've also stepped through in debug and can see the step being marked as failed but I can't see that the screenshot is ever being taken...

**Comment by c-fau** [comment]
For simplicity I adjusted one of the feature files to only include a single step which fails (not pushed to the repo), when this is ran and the step has failed, for some reason the status is appearing as 'PASSED' when left to run, no screenshot is taken and nothing is attached to the report, when manually set to 'FAILED' the screenshot is taken and appears in the report.
Where in the code does the step result get set, is it potentially getting set and then overwritten?

<img width="1708" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/41241f26-42ef-42fa-bf78-1c4bb22a1572">

<img width="1020" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/263dd053-3577-4d09-861f-fdadf2776f24">

**Comment by c-fau** [comment]
I have pulled and built the project in this repo with gradle and can confirm the issue is also present. No screenshot is produced, the step is marked as 'SUCCESS'.
https://github.com/serenity-bdd/serenity-cucumber-page-objects-starter/tree/master
<img width="1713" alt="image" src="https://github.com/serenity-bdd/serenity-core/assets/124570901/cc189ba6-4544-469d-a626-43e6b6150f91">

**Comment by wakaleo** [comment]
OK, the reason it doesn't work is because with parallel execution, we don't/can't know the status of the test when that code is called. This happens when the assertions are made at the step definition level (i.e. when not using Screenplay). It will be hard to fix.

**Comment by c-fau** [comment]
I had a play a few days ago and creating a class to manage the "actions" inbetween the step definitions and the page objects does allow a screenshot to be taken on failure, not using screenplay, i've also updated the linked repo to include this.
This is the flow of the classes.
feature file > stepDefs > pageActions > pageObjects

Closing this ticket as resolved.

**Comment by Johan-Sap** [comment]
@wakaleo and @c-fau I encountered this issue as well lately. I dove into the code and came to similar conclusions but also managed to fix it locally. Instead of checking if the screenshots needs to be taken in the method '`takeEndOfStepScreenshotForRecording`' I just make the screenshot **ALLWAYS** (like FOR_EACH_ACTION). Then in the '`takeEndOfStepScreenshotForPlayback`' I check if the screenshots needs to be added or not using the method I removed from the '....ForRecording' method.

This left me with cleaning up any screenshots that were taken while they are not used in the report. If this would not be done, there would be screenshots in the outputDir (target/serenity/) that do not show up in the report. Doing this required me to track which screenshots were actually used in the end. I created a method '`removeUnnecesaryScreenshotsFromOutputDir`' in the BaseStepListener to manage this part. 

Calling this method happens in the '**stepFailed**' and '**testFinished**' method also in the BaseStepListener. 

I can however not determine the impact on other test structures like screenplay for instance. I will make a PR when 4.0.30 version is out and base my fix on that version.

**Comment by wakaleo** [comment]
4.0.30 is now out

**Comment by Johan-Sap** [comment]
Created a PR for the fix. Please look at is because (like mentioned) I find it difficult to determine impact on the way screenplay for instance works. https://github.com/serenity-bdd/serenity-core/pull/3361

Locally it all works as expected. Behaviour of other screenshot levels is the same as in 4.0.30.
